pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/ContextImpl.java

```diff
@@ -480,27 +481,54 @@ public void recordMetric(String metricName, double value) {
 
         if (producer == null) {
 
-            Producer<O> newProducer = ((ProducerBuilderImpl<O>) pulsar.getProducerBuilder().clone())
-                    .schema(schema)
-                    .blockIfQueueFull(true)
-                    .enableBatching(true)
-                    .batchingMaxPublishDelay(10, TimeUnit.MILLISECONDS)
-                    .compressionType(CompressionType.LZ4)
-                    .hashingScheme(HashingScheme.Murmur3_32Hash) //
-                    .messageRoutingMode(MessageRoutingMode.CustomPartition)
-                    .messageRouter(FunctionResultRouter.of())
-                    // set send timeout to be infinity to prevent potential deadlock with consumer
-                    // that might happen when consumer is blocked due to unacked messages
-                    .sendTimeout(0, TimeUnit.SECONDS)
-                    .topic(topicName)
-                    .properties(InstanceUtils.getProperties(componentType,
-                            FunctionCommon.getFullyQualifiedName(
-                                    this.config.getFunctionDetails().getTenant(),
-                                    this.config.getFunctionDetails().getNamespace(),
-                                    this.config.getFunctionDetails().getName()),
-                            this.config.getInstanceId()))
-                    .create();
-
+            Producer<O> newProducer = null;
+
+            if (this.config.getFunctionDetails() != null && this.config.getFunctionDetails().getSink() != null &&
+            this.config.getFunctionDetails().getSink().getProducerSpec() != null){
```

Comment:
`this.config.getFunctionDetails().getSink().getProducerSpec()` will never return null.  This is technically changing the existing behavior of `publish` when called from `Context`.  I don't think this is the right change. If a user want to use a custom producer to produce in the function (instead of just returning), the user should be able to:

1. Instantiate a PulsarClient, create a producer, and sent messages (already can do this today)
2. We can add a new method called "newProducerBuilder()" to the Context in which the returned builder automatically inherits some of the settings such as authentication from the function.

Reply:
@jerrypeng Thanks for reviewing the PR. I want to better understand your suggestion. I'm hearing several things, so please correct any of my misunderstandings:
1. Instead of performing the null check on `this.config.getFunctionDetails().getSink().getProducerSpec()`, I could skip the null check and just pass through the protobuf properties directly. The defaults they use match the existing configurations, but it sounds like there's a concern that it could still change existing behavior? 
2. You're saying that I could define a new method, `newProducerBuilder()` to inherit the protobuf settings (instead of passing them through directly, like I mentioned above), but it sounds like there are other settings I'd want to inherit as well (such as relating to authentication)? 

Would the `newProducerBuilder()` method be used instead of `getProducer()` by functions that want to use the new settings? It sounds like that would require a code change for any existing function that we want to update to use the new settings. Is that right?